File: solutions/java/twelve-days/src/main/java/TwelveDays.java
import java.util.StringJoiner;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class TwelveDays {
    private static final String[] DAYS = {
        "first", "second", "third", "fourth", "fifth", "sixth",
        "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"
    };
    private static final String[] GIFTS = {
        "a Partridge in a Pear Tree",
        "two Turtle Doves, and",
        "three French Hens,",
        "four Calling Birds,",
        "five Gold Rings,",
        "six Geese-a-Laying,",
        "seven Swans-a-Swimming,",
        "eight Maids-a-Milking,",
        "nine Ladies Dancing,",
        "ten Lords-a-Leaping,",
        "eleven Pipers Piping,",
        "twelve Drummers Drumming,"
    };

    String verse(int verseNumber) {
        var i = verseNumber - 1;
        var song = new StringJoiner(" ",
            "On the " + DAYS[i] + " day of Christmas my true love gave to me: ",
            ".\n");

        for (; i >= 0; --i) {
            song.add(GIFTS[i]);
        }
        return song.toString();
    }

    String verses(int startVerse, int endVerse) {
        return IntStream
            .rangeClosed(startVerse, endVerse)
            .mapToObj(this::verse)
            .collect(Collectors.joining("\n"));
    }

    String sing() {
        return verses(1, 12);
    }
}

File: solutions/python/twelve-days/twelve_days.py
days = (
    ("first", "a Partridge in a Pear Tree."),
    ("second", "two Turtle Doves, and"),
    ("third", "three French Hens,"),
    ("fourth", "four Calling Birds,"),
    ("fifth", "five Gold Rings,"),
    ("sixth", "six Geese-a-Laying,"),
    ("seventh", "seven Swans-a-Swimming,"),
    ("eighth", "eight Maids-a-Milking,"),
    ("ninth", "nine Ladies Dancing,"),
    ("tenth", "ten Lords-a-Leaping,"),
    ("eleventh", "eleven Pipers Piping,"),
    ("twelfth", "twelve Drummers Drumming,"),
)


def recite_single_verse(verse_number: int) -> str:
    day = days[verse_number - 1][0]
    gifts = " ".join(days[i][1] for i in reversed(range(verse_number)))
    return f"On the {day} day of Christmas my true love gave to me: {gifts}"


def recite(start_verse, end_verse):
    verses = range(start_verse, end_verse + 1)
    return list(map(recite_single_verse, verses))
